sbclPackages.cl-autowrap: compile cl-autowrap/libffi

robgssp opened the pull request:
## Description of changes

To do so, the system is added and libffi native dep is included.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

robgssp committed:
sbclPackages.cl-autowrap: compile cl-autowrap/libffi

include libffi dependecy for this

7c6f434c commented:
@ofborg build sbcl.pkgs.cl-autowrap

7c6f434c commented:
Oh, those two native-library-adding PRs are in conflict…

Anyway, should we restrict platforms here?

Uthar commented:
Ah, we've found a circular dependency! `cl-plus-c.asd` <-> `cl-autowrap.asd`. I wonder how will `sdl2` behave - it depends on both!  There could be exciting behavior when someone overrides one but not the other. On the other hand, merging them into a single "jumbo" package is not ideal, because some packages depend on just one. At the same time, the jumbo variant would conflict with the base variants in the source registry. I wonder how does Guix do it?

7c6f434c commented:
I remember that most of the time such dependency are just for tests (in one of the directions), maybe here we could also cut the tests and pretend things are good?

robgssp commented:
Rebased to latest.

I'd argue that merging the packages makes sense, as `cl-plus-c` is a) in the same repo, and b) doesn't include any additional dependencies over `cl-autowrap`. The more questionable inclusion is wrapping `cl-autowrap` and `cl-autowrap/libffi` together, as this means that systems that don't actually use libffi still pull in the dependency. That said, I'd like to leave it as-is since no other lisp package here exposes a slashed system separately afaict. Guix, fwiw, bundles `cl-autowrap` and `cl-plus-c` and doesn't include `/libffi`.

`sdl2` is tested working with the change.

As far as overriding the package but not the alias goes, you're right that it's not ideal but also no worse than the status quo - cl-plus-c is currently a separately-specified build of the same repo in imported.nix.

@7c6f434c so in this case `cl-plus-c` isn't just tests, but it is a closely-related library to cl-autowrap so I think it does make sense to bundle their builds together.

7c6f434c commented:
Same repo sounds like a good argument for merging to me

7c6f434c commented:
@ofborg build sbcl.pkgs.cl-autowrap

7c6f434c commented:
Does it also want c2ffi (for aarch64-linux)?

robgssp commented:
Sure does! Declaring a `propagatedBuildInput` works, better would be substituting the default `autowrap:*c2ffi-program*` with a static path, but nix's asdf build process makes that weirdly tricky. Substituting in `postConfigurePhase` does not have the desired effect since the build ends up referring to an external source derivation instead. I'll poke at it more later.

Uthar commented:
hmm yes, it's compiling from src instead of from pwd so that "go-to-definition" works. I didn't know how to do that differently. I saw that SBCL does have the required parameter to with-compilation-unit, but didn't dig into others. Maybe we can fix it.

For now you could use patches, postPatch to alter sources

robgssp committed:
sbclPackages.cl-autowrap: bake in c2ffi path

robgssp commented:
Added fixing the c2ffi path.

7c6f434c commented:
@ofborg build sbcl.pkgs.cl-autowrap

7c6f434c commented:
Hm OK, now this thing on aarch64-linux sounds like something without a completely straightforward fix… Should I merge the current state as an improvement at least on x86-84?

robgssp commented:
Let's not, cl-autowrap does currently build on aarch64 so this'll break anyone that uses it there. I'll take a look at the failure.

robgssp committed:
c2ffi: add wrapper script and setup hook

c2ffi needs to know include paths similar to cc's, so add a wrapper
and hook similar to cc-wrapper to collect them.

7c6f434c commented on pkgs/development/tools/misc/c2ffi/c2ffi-wrapper.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+exec @c2ffi@ $@ -I . -i @libc@/include -i @clang@/resource-root/include $NIX_C2FFI_FLAGS
```
Does this do anything that `wrapProgram` would not be able to provide?
